File: mkdocs/snippets/devbook/transactions/bonded_aggregate.mjs
import { Hash256, PrivateKey } from 'symbol-sdk';
import {
	SymbolFacade,
	descriptors,
	generateMosaicAliasId,
	models
} from 'symbol-sdk/symbol';

const NODE_URL = process.env.NODE_URL ||
	'https://reference.symboltest.net:3001';
console.log('Using node', NODE_URL);

// Helper function to announce transaction
async function announceTransaction(payload, endpoint, label) {
	console.log(`Announcing ${label} to ${endpoint}`);
	const response = await fetch(`${NODE_URL}${endpoint}`, {
		method: 'PUT',
		headers: { 'Content-Type': 'application/json' },
		body: payload
	});
	console.log('  Response:', await response.text());
}

// Helper function to wait for transaction status
async function waitForStatus(hash, expectedStatus, label) {
	console.log(
		`Waiting for ${label} to reach ${expectedStatus} status...`);
	let attempts = 0;
	const maxAttempts = 60;

	while (attempts < maxAttempts) {
		try {
			const url = `${NODE_URL}/transactionStatus/${hash}`;
			const response = await fetch(url);

			if (!response.ok) {
				const error = new Error(
					`HTTP ${response.status}: ${response.statusText}`);
				error.status = response.status;
				throw error;
			}

			const status = await response.json();

			console.log('  Transaction status:', status.group);

			if ('failed' === status.group)
				throw new Error(`${label} failed: ${status.code}`);

			if (status.group === expectedStatus) {
				console.log(
					`${label} ${expectedStatus} in ${attempts} seconds`
				);
				return;
			}
		} catch (error) {
			if (404 === error.status)
				console.log('  Transaction status: not yet available');
			else
				throw error;
		}

		attempts++;
		await new Promise(resolve => { setTimeout(resolve, 1000); });
	}

	throw new Error(
		`${label} not ${expectedStatus} after ${maxAttempts} attempts`
	);
}

// Account A (initiates the aggregate tx and sends XYM to Account B) [>step-1]
const ACCOUNT_A_PRIVATE_KEY = process.env.ACCOUNT_A_PRIVATE_KEY || (
	'0000000000000000000000000000000000000000000000000000000000000000');
const accountAKeyPair = new SymbolFacade.KeyPair(
	new PrivateKey(ACCOUNT_A_PRIVATE_KEY));

// Account B (sends custom mosaic to Account A)
const ACCOUNT_B_PRIVATE_KEY = process.env.ACCOUNT_B_PRIVATE_KEY || (
	'1111111111111111111111111111111111111111111111111111111111111111');
const accountBKeyPair = new SymbolFacade.KeyPair(
	new PrivateKey(ACCOUNT_B_PRIVATE_KEY));

const facade = new SymbolFacade('testnet');
const accountAAddress = facade.network.publicKeyToAddress(
	accountAKeyPair.publicKey);
const accountBAddress = facade.network.publicKeyToAddress(
	accountBKeyPair.publicKey);
console.log('Account A:', accountAAddress.toString());
console.log('Account B:', accountBAddress.toString());
// [<step-1]
try {
	// Fetch recommended fees [>step-2]
	const feePath = '/network/fees/transaction';
	console.log('Fetching recommended fees from', feePath);
	const feeResponse = await fetch(`${NODE_URL}${feePath}`);
	const feeJSON = await feeResponse.json();
	const medianMultiplier = feeJSON.medianFeeMultiplier;
	const minimumMultiplier = feeJSON.minFeeMultiplier;
	const feeMultiplier = Math.max(medianMultiplier, minimumMultiplier);
	console.log('  Fee multiplier:', feeMultiplier); // [<step-2]

	// Embedded tx 1: Account A transfers 10 XYM to Account B [>step-3]
	const embeddedTransaction1 =
		facade.createEmbeddedTransactionFromTypedDescriptor(
			new descriptors.TransferTransactionV1Descriptor(
				accountBAddress,
				[
					new descriptors.UnresolvedMosaicDescriptor(
						generateMosaicAliasId('symbol.xym'),
						new models.Amount(10_000_000n)) // 10 XYM
				],
				undefined),
			accountAKeyPair.publicKey);

	// Embedded tx 2: Account B transfers 1 custom mosaic to Account A
	const customMosaicId = 0x6D1314BE751B62C2n;
	const embeddedTransaction2 =
		facade.createEmbeddedTransactionFromTypedDescriptor(
			new descriptors.TransferTransactionV1Descriptor(
				accountAAddress,
				[
					new descriptors.UnresolvedMosaicDescriptor(
						customMosaicId,
						new models.Amount(1n)) // 1 custom mosaic
				],
				undefined),
			accountBKeyPair.publicKey); // [<step-3]

	// Build the bonded aggregate transaction [>step-4]
	const embeddedTransactions = [
		embeddedTransaction1, embeddedTransaction2];
	const bondedTransaction = facade.createTransactionFromTypedDescriptor(
		new descriptors.AggregateBondedTransactionV3Descriptor(
			facade.static.hashEmbeddedTransactions(embeddedTransactions),
			embeddedTransactions,
			undefined),
		accountAKeyPair.publicKey,
		feeMultiplier,
		2 * 60 * 60,
		1);
	console.log('Built aggregate without signatures:');
	console.log(JSON.stringify(bondedTransaction.toJson(), null, 2));
	// [<step-4]
	// --- ACCOUNT A (Initiator) --- [>step-5]
	// Sign the bonded aggregate transaction
	console.log('[Account A] Signing the bonded aggregate...');
	const bondedSignature = facade.signTransaction(
		accountAKeyPair, bondedTransaction);
	const bondedJsonPayload = facade.transactionFactory.static
		.attachSignature(bondedTransaction, bondedSignature);
	const bondedHash = facade.hashTransaction(
		bondedTransaction).toString();
	console.log('Bonded aggregate transaction hash:', bondedHash);
	// [<step-5]
	// Create hash lock transaction [>step-6]
	console.log('Creating hash lock transaction...');
	const hashLock = facade.createTransactionFromTypedDescriptor(
		new descriptors.HashLockTransactionV1Descriptor(
			new descriptors.UnresolvedMosaicDescriptor(
				generateMosaicAliasId('symbol.xym'),
				new models.Amount(10_000_000n)), // 10 XYM deposit
			new models.BlockDuration(100n),
			new Hash256(bondedHash)),
		accountAKeyPair.publicKey,
		feeMultiplier,
		2 * 60 * 60);

	// Sign hash lock
	console.log('[Account A] Signing the hash lock...');
	const hashLockSignature = facade.signTransaction(
		accountAKeyPair, hashLock);
	const hashLockPayload = facade.transactionFactory.static
		.attachSignature(hashLock, hashLockSignature);
	const hashLockHash = facade.hashTransaction(hashLock)
		.toString();
	console.log('Hash lock transaction hash:', hashLockHash);

	// Announce hash lock and wait for confirmation
	await announceTransaction(
		hashLockPayload, '/transactions', 'Hash lock'
	);
	await waitForStatus(hashLockHash, 'confirmed', 'Hash lock');
	// [<step-6]
	// Announce bonded aggregate and wait for partial status // [>step-7]
	await announceTransaction(
		bondedJsonPayload, '/transactions/partial',
		'Bonded aggregate transaction'
	);
	await waitForStatus(
		bondedHash, 'partial', 'Bonded aggregate transaction'
	); // [<step-7]

	// --- ACCOUNT B (Cosigner) --- [>step-8]
	// Retrieves partial transactions waiting for signature
	const partialPath =
		`/transactions/partial?address=${accountBAddress}`;
	console.log(
		'[Account B] Checking for partial transactions from ' +
		'/transactions/partial'
	);
	const partialResponse = await fetch(`${NODE_URL}${partialPath}`);
	const partialTxs = await partialResponse.json();
	if (!partialTxs.data || 0 === partialTxs.data.length)
		throw new Error('No partial transactions found');

	console.log(
		`Found ${partialTxs.data.length} partial transaction(s)`);

	// Find the transaction matching the expected hash
	const found = partialTxs.data.some(
		tx => tx.meta.hash === bondedHash
	);
	if (!found) {
		throw new Error(
			`Expected transaction ${bondedHash} not found in ` +
			'partial transactions'
		);
	}
	console.log(`Found matching transaction: ${bondedHash}`);
	// [<step-8]
	// Fetch full transaction details using the hash [>step-9]
	const detailPath = `/transactions/partial/${bondedHash}`;
	const detailResponse = await fetch(`${NODE_URL}${detailPath}`);
	const partialTxJson = await detailResponse.json();

	// Verify transaction content before cosigning
	const txData = partialTxJson.transaction;
	console.log('[Account B] Verifying transaction: ' +
		`${txData.transactions.length} embedded transactions`
	); // [<step-9]

	// Submit Account B's cosignature using the transaction hash [>step-10]
	const cosignaturePath = '/transactions/cosignature';
	console.log('[Account B] Cosigning the bonded aggregate...');
	const cosignature = SymbolFacade.cosignTransactionHash(
		accountBKeyPair, new Hash256(bondedHash), true
	);
	const cosignaturePayload = JSON.stringify({
		version: cosignature.version.toString(),
		signerPublicKey: cosignature.signerPublicKey.toString(),
		signature: cosignature.signature.toString(),
		parentHash: cosignature.parentHash.toString()
	});

	// Announce cosignature
	await announceTransaction(
		cosignaturePayload, cosignaturePath, 'cosignature'
	); // [<step-10]

	// Wait for final confirmation [>step-11]
	await waitForStatus(
		new Hash256(bondedHash), 'confirmed',
		'Bonded aggregate transaction'
	); // [<step-11]
} catch (e) {
	console.error(e.message, '| Cause:', e.cause?.code ?? 'unknown');
}

File: mkdocs/snippets/devbook/transactions/BondedAggregate.java
//JAVA 21+
//DEPS org.symbol:symbol-sdk:3.3.1

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandlers;
import java.util.List;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.symbol.sdk.CryptoTypes;
import org.symbol.sdk.facade.SymbolFacade;
import org.symbol.sdk.symbol.Address;
import org.symbol.sdk.symbol.IdGenerator;
import org.symbol.sdk.symbol.KeyPair;
import org.symbol.sdk.symbol.SymbolTransactionFactory;
import org.symbol.sdk.symbol.descriptors.*;
import org.symbol.sdk.symbol.models.*;

public final class BondedAggregate {
	private static final ObjectMapper JSON_MAPPER = new ObjectMapper();

	private static final HttpClient HTTP_CLIENT =
		HttpClient.newHttpClient();

	private final String nodeUrl = System.getenv().getOrDefault(
		"NODE_URL", "https://reference.symboltest.net:3001");

	private final SymbolFacade facade = new SymbolFacade("testnet");

	private KeyPair accountAKeyPair;

	private KeyPair accountBKeyPair;

	private Address accountAAddress;

	private Address accountBAddress;

	// Helper function to announce transaction
	private void announceTransaction(
		final String payload,
		final String endpoint,
		final String label
	) throws IOException, InterruptedException {
		System.out.printf("Announcing %s to %s%n", label, endpoint);
		final HttpRequest request = HttpRequest.newBuilder(
			URI.create(nodeUrl + endpoint))
			.header("Content-Type", "application/json")
			.PUT(HttpRequest.BodyPublishers.ofString(payload))
			.build();
		final HttpResponse<String> response = HTTP_CLIENT.send(
			request, BodyHandlers.ofString());
		System.out.printf("  Response: %s%n", response.body());
	}

	// Helper function to wait for transaction status
	private void waitForStatus(
		final String hash,
		final String expectedStatus,
		final String label
	) throws IOException, InterruptedException {
		System.out.printf(
			"Waiting for %s to reach %s status...%n",
			label, expectedStatus);
		int attempts = 0;
		final int maxAttempts = 60;

		while (attempts < maxAttempts) {
			final String url = nodeUrl + "/transactionStatus/" + hash;
			final HttpRequest request = HttpRequest.newBuilder(
				URI.create(url)).GET().build();
			final HttpResponse<String> response = HTTP_CLIENT.send(
				request, BodyHandlers.ofString());

			if (response.statusCode() / 100 != 2) {
				if (404 == response.statusCode()) {
					System.out.println(
						"  Transaction status: not yet available");
				} else
					throw new IOException(
						"HTTP " + response.statusCode());
			} else {
				final JsonNode status = JSON_MAPPER.readTree(
					response.body());

				System.out.printf("  Transaction status: %s%n",
					status.get("group").asText());

				if ("failed".equals(status.get("group").asText()))
					throw new IOException(String.format("%s failed: %s",
						label, status.get("code").asText()));

				if (status.get("group").asText().equals(expectedStatus)) {
					System.out.printf("%s %s in %d seconds%n",
						label, expectedStatus, attempts);
					return;
				}
			}

			++attempts;
			Thread.sleep(1000);
		}

		throw new IOException(String.format(
			"%s not %s after %d attempts",
			label, expectedStatus, maxAttempts));
	}

	public static void main(final String[] args) {
		try {
			new BondedAggregate().run();
		} catch (final Exception ex) {
			System.out.println(null == ex.getMessage()
				? ex.toString()
				: ex.getMessage());
		}
	}

	private void run() throws IOException, InterruptedException {
		System.out.printf("Using node %s%n", nodeUrl);

		// [>step-1]
		// Account A (initiates the aggregate tx
		// and sends XYM to Account B)
		final String accountAPrivateKey = System.getenv().getOrDefault(
			"ACCOUNT_A_PRIVATE_KEY", "0".repeat(64));
		accountAKeyPair = new KeyPair(
			new CryptoTypes.PrivateKey(accountAPrivateKey));

		// Account B (sends custom mosaic to Account A)
		final String accountBPrivateKey = System.getenv().getOrDefault(
			"ACCOUNT_B_PRIVATE_KEY", "1".repeat(64));
		accountBKeyPair = new KeyPair(
			new CryptoTypes.PrivateKey(accountBPrivateKey));

		accountAAddress = facade.network.publicKeyToAddress(
			accountAKeyPair.getPublicKey());
		accountBAddress = facade.network.publicKeyToAddress(
			accountBKeyPair.getPublicKey());
		System.out.printf("Account A: %s%n", accountAAddress);
		System.out.printf("Account B: %s%n", accountBAddress);
		// [<step-1]

		// Fetch recommended fees [>step-2]
		final String feePath = "/network/fees/transaction";
		System.out.printf(
			"Fetching recommended fees from %s%n", feePath);
		final HttpRequest feeRequest = HttpRequest.newBuilder(
			URI.create(nodeUrl + feePath)).GET().build();
		final HttpResponse<String> feeResponse = HTTP_CLIENT.send(
			feeRequest, BodyHandlers.ofString());
		final JsonNode feeJSON = JSON_MAPPER.readTree(
			feeResponse.body());
		final long medianMultiplier =
			feeJSON.get("medianFeeMultiplier").asLong();
		final long minimumMultiplier =
			feeJSON.get("minFeeMultiplier").asLong();
		final long feeMultiplier = Math.max(
			medianMultiplier, minimumMultiplier);
		System.out.printf("  Fee multiplier: %d%n", feeMultiplier);
		// [<step-2]
		// Embedded tx 1: Account A transfers 10 XYM to Account B [>step-3]
		final EmbeddedTransaction embeddedTransaction1 =
			facade.createEmbeddedTransactionFromTypedDescriptor(
				new TransferTransactionV1Descriptor(
					accountBAddress,
					List.of(new UnresolvedMosaicDescriptor(
						new UnresolvedMosaicId(
							IdGenerator.generateMosaicAliasId(
								"symbol.xym")),
							// 10 XYM
							new Amount(10_000_000))),
					null),
				accountAKeyPair.getPublicKey());

		// Embedded tx 2: Account B transfers 1 custom mosaic to Account A
		final long customMosaicId = 0x6D1314BE751B62C2L;
		final EmbeddedTransaction embeddedTransaction2 =
			facade.createEmbeddedTransactionFromTypedDescriptor(
				new TransferTransactionV1Descriptor(
					accountAAddress,
					List.of(new UnresolvedMosaicDescriptor(
						new UnresolvedMosaicId(customMosaicId),
							// 1 custom mosaic
							new Amount(1))),
					null),
				accountBKeyPair.getPublicKey());
		// [<step-3]
		// Build the bonded aggregate transaction [>step-4]
		final List<EmbeddedTransaction> embeddedTransactions =
			List.of(embeddedTransaction1, embeddedTransaction2);
		final Transaction bondedTransaction =
			facade.createTransactionFromTypedDescriptor(
				new AggregateBondedTransactionV3Descriptor(
					SymbolFacade.hashEmbeddedTransactions(
						embeddedTransactions),
					embeddedTransactions,
					null),
				accountAKeyPair.getPublicKey(),
				feeMultiplier,
				2 * 60 * 60,
				1);
		System.out.println("Built aggregate without signatures:");
		System.out.println(JSON_MAPPER.writerWithDefaultPrettyPrinter()
			.writeValueAsString(bondedTransaction.toJson()));
		// [<step-4]
		// --- ACCOUNT A (Initiator) --- [>step-5]
		// Sign the bonded aggregate transaction
		System.out.println("[Account A] Signing the bonded aggregate...");
		final CryptoTypes.Signature bondedSignature =
			facade.signTransaction(accountAKeyPair, bondedTransaction);
		final String bondedJsonPayload = SymbolTransactionFactory
			.attachSignature(bondedTransaction, bondedSignature);
		final String bondedHash = facade.hashTransaction(
			bondedTransaction).toString();
		System.out.printf("Bonded aggregate transaction hash: %s%n",
			bondedHash);
		// [<step-5]
		// Create hash lock transaction [>step-6]
		System.out.println("Creating hash lock transaction...");
		final Transaction hashLock =
			facade.createTransactionFromTypedDescriptor(
				new HashLockTransactionV1Descriptor(
					new UnresolvedMosaicDescriptor(
						new UnresolvedMosaicId(
							IdGenerator.generateMosaicAliasId(
								"symbol.xym")),
						new Amount(10_000_000)), // 10 XYM deposit
					new BlockDuration(100),
					new CryptoTypes.Hash256(bondedHash)),
				accountAKeyPair.getPublicKey(),
				feeMultiplier,
				2 * 60 * 60);

		// Sign hash lock
		System.out.println("[Account A] Signing the hash lock...");
		final CryptoTypes.Signature hashLockSignature =
			facade.signTransaction(accountAKeyPair, hashLock);
		final String hashLockPayload = SymbolTransactionFactory
			.attachSignature(hashLock, hashLockSignature);
		final String hashLockHash = facade.hashTransaction(hashLock)
			.toString();
		System.out.printf("Hash lock transaction hash: %s%n",
			hashLockHash);

		// Announce hash lock and wait for confirmation
		announceTransaction(
			hashLockPayload, "/transactions", "Hash lock"
		);
		waitForStatus(hashLockHash, "confirmed", "Hash lock");
		// [<step-6]
		// Announce bonded aggregate and wait for partial status [>step-7]
		announceTransaction(
			bondedJsonPayload, "/transactions/partial",
			"Bonded aggregate transaction"
		);
		waitForStatus(
			bondedHash, "partial", "Bonded aggregate transaction"
		);
		// [<step-7]
		// --- ACCOUNT B (Cosigner) --- [>step-8]
		// Retrieves partial transactions waiting for signature
		final String partialPath =
			"/transactions/partial?address=" + accountBAddress;
		System.out.println(
			"[Account B] Checking for partial transactions from "
			+ "/transactions/partial"
		);
		final HttpRequest partialRequest = HttpRequest.newBuilder(
			URI.create(nodeUrl + partialPath)).GET().build();
		final HttpResponse<String> partialResponse = HTTP_CLIENT.send(
			partialRequest, BodyHandlers.ofString());
		final JsonNode partialTxs = JSON_MAPPER.readTree(
			partialResponse.body());
		if (!partialTxs.has("data") || partialTxs.get("data").isEmpty())
			throw new IOException("No partial transactions found");

		System.out.printf("Found %d partial transaction(s)%n",
			partialTxs.get("data").size());

		// Find the transaction matching the expected hash
		boolean found = false;
		for (final JsonNode tx : partialTxs.get("data"))
				found |= bondedHash.equals(
					tx.get("meta").get("hash").asText());
		if (!found) {
			throw new IOException(
				"Expected transaction " + bondedHash
				+ " not found in partial transactions"
			);
		}
		System.out.printf("Found matching transaction: %s%n", bondedHash);
		// [<step-8]
		// Fetch full transaction details using the hash [>step-9]
		final String detailPath = "/transactions/partial/" + bondedHash;
		final HttpRequest detailRequest = HttpRequest.newBuilder(
			URI.create(nodeUrl + detailPath)).GET().build();
		final HttpResponse<String> detailResponse = HTTP_CLIENT.send(
			detailRequest, BodyHandlers.ofString());
		final JsonNode partialTxJson = JSON_MAPPER.readTree(
			detailResponse.body());

		// Verify transaction content before cosigning
		final JsonNode txData = partialTxJson.get("transaction");
		System.out.println("[Account B] Verifying transaction: "
			+ txData.get("transactions").size()
			+ " embedded transactions"
		);
		// [<step-9]
		// [>step-10]
		// Submit Account B's cosignature using the transaction hash
		final String cosignaturePath = "/transactions/cosignature";
		System.out.println(
			"[Account B] Cosigning the bonded aggregate...");
		final DetachedCosignature cosignature =
			SymbolFacade.cosignTransactionHashDetached(
				accountBKeyPair, new CryptoTypes.Hash256(bondedHash)
			);
		final String cosignaturePayload =
			JSON_MAPPER.writeValueAsString(cosignature.toJson());

		// Announce cosignature
		announceTransaction(
			cosignaturePayload, cosignaturePath, "cosignature"
		);
		// [<step-10]
		// Wait for final confirmation [>step-11]
		waitForStatus(
			bondedHash, "confirmed",
			"Bonded aggregate transaction"
		); // [<step-11]
	}
}
